src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -214,7 +214,7 @@ Pipe MergeTreeDataSelectExecutor::readFromParts(
 
     KeyCondition key_condition(query_info, context, primary_key_columns, primary_key.expression);
 
-    if (settings.force_primary_key && key_condition.alwaysUnknownOrTrue())
+    if (settings.force_primary_key && key_condition.indexCannotBeUsed(true))
```

Comment:
Now we have multiple callsites with `indexCannotBeUsed(true)` and this will be confusing for readers. It's better to have different names for these two variants.

`alwaysUnknownOrTrue` conveys the message very well.

For partition pruning, we need `anyUnknownOrAlwaysTrue` .

Reply:
>Now we have multiple callsites with indexCannotBeUsed(true) and this will be confusing for readers. It's better to have different names for these two variants.

Indeed, I kind of didn't like it either, but did not find better naming to distinguish those two

>For partition pruning, we need anyUnknownOrAlwaysTrue .

OK, renamed (and rebased against upstream/master that includes you PartitionPruner)